### ISSUE TITLE
Towards Natural Language Semantic Code Search

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#54]
Towards Natural Language Semantic Code Search<br>
<br>
https://ift.tt/2MK4O97<br>
<br>
hamelsmu<br>
<br>
<div>
<div>
<p><img src="https://githubengineering.com/images/semantic_code_search/main.jpg" alt="Hubot"></p>
<h2 id="motivation">Motivation</h2>
<p>Searching code on GitHub is currently limited to keyword search. This assumes either the user knows the syntax, or can anticipate what keywords might be in comments surrounding the code they are looking for. Our machine learning scientists have been researching ways to enable the <a href="https://en.wikipedia.org/wiki/Semantic_search">semantic search</a> of code.</p>
<p>To fully grasp the concept of semantic search, consider the below search query, <em>&ldquo;ping REST api and return results&rdquo;</em>:</p>
<p><img src="https://githubengineering.com/images/semantic_code_search/hubot_codesearch.gif" alt="Vector-space diagram"></p>
<p>Note that the demonstrated semantic search returns reasonable results even though there are no keywords in common between the search query and the text (the code &amp; comments found do not contain the words &ldquo;Ping&rdquo;, &ldquo;REST&rdquo; or &ldquo;api&rdquo;)! The implications of augmenting keyword search with semantic search are profound. For example, such a capability would expedite the process of on-boarding new software engineers onto projects and bolster the discoverability of code in general.</p>
<p>In this post, we want to share how we are leveraging deep learning to make progress towards this goal. <em>We also share an open source example with code and data that you can use to reproduce these results!</em></p>
<h2 id="introduction">Introduction</h2>
<p>One of the key areas of machine learning research underway at GitHub is representation learning of entities, such as repos, code, issues, profiles and users. We have made significant progress towards enabling semantic search by learning representations of code that share a common vector space as text. For example, consider the below diagram:</p>
<p><img src="https://githubengineering.com/images/semantic_code_search/shared_vecs.png" alt="Vector-space diagram"></p>
<p>In the above example, Text 2 (blue) is a reasonable description of the code, whereas Text 1 (red) is not related to the code at all. Our goal is to learn representations where (text, code) pairs that describe the same concept are close neighbors, whereas unrelated (text, code) pairs are further apart. By representing text and code in the same vector space, we can vectorize a user&rsquo;s search query and lookup the nearest neighbor that represents code. Below is a four-part description of the approach we are currently using to accomplish this task:</p>
<h3 id="1-learning-representations-of-code">1. Learning Representations of Code</h3>
<p>In order to learn a representation of code, we train a <a href="https://towardsdatascience.com/how-to-create-data-products-that-are-magical-using-sequence-to-sequence-models-703f86a231f8">sequence-to-sequence model</a> that learns to summarize code. A way to accomplish this for Python is to supply (code, docstring) pairs where the docstring is the target variable the model is trying to predict. One active area of research for us is incorporating domain specific optimizations like <a href="https://arxiv.org/pdf/1802.00921.pdf">tree-based LSTMs</a>, <a href="https://github.com/Microsoft/gated-graph-neural-network-samples">gated-graph networks</a> and syntax-aware tokenization. Below is a screenshot that showcases the code summarizer model at work. In this example, there are two python functions supplied as input, and in both cases the model produces a reasonable summary of the code as output:</p>
<p><img src="https://githubengineering.com/images/semantic_code_search/code_summarizer.png" alt="code summarizer"></p>
<p>It should be noted that in the above examples, the model produces the summary by using the entire code blob, not merely the function name.</p>
<p>Building a code summarizer is a very exciting project on its own, however, we can utilize the encoder from this model as a general purpose feature extractor for code. After extracting the encoder from this model, we can <a href="https://flyyufelix.github.io/2016/10/03/fine-tuning-in-keras-part1.html">fine-tune</a> it for the task of mapping code to the vector space of natural language.</p>
<p>We can evaluate this model objectively using the <a href="https://en.wikipedia.org/wiki/BLEU">BLEU score</a>. Currently we have been able to achieve a BLEU score of 13.5 on a holdout set of python code, using the <a href="https://github.com/pytorch/fairseq">fairseq-py library</a> for sequence to sequence models.</p>
<h3 id="2-learning-representations-of-text-phrases">2. Learning Representations of Text Phrases</h3>
<p>In addition to learning a representation for code, we needed to find a suitable representation for short phrases (like sentences found in Python docstrings). Initially, we experimented with the <a href="https://arxiv.org/abs/1803.11175">Universal Sentence Encoder</a>, a pre-trained encoder for text that is <a href="https://www.tensorflow.org/hub/modules/google/universal-sentence-encoder/1">available on TensorFlow Hub</a>. While the embeddings from worked reasonably well, we found that it was advantageous to learn embeddings that were specific to the vocabulary and semantics of software development. One area of ongoing research involves evaluating different domain-specific corpuses for training our own model, ranging from GitHub issues to third party datasets.</p>
<p>To learn this representation of phrases, we trained a <a href="https://en.wikipedia.org/wiki/Language_model">neural language model</a> by leveraging the <a href="https://github.com/fastai/fastai">fast.ai</a> library. This library gave us easy access to state of the art architectures such as <a href="https://arxiv.org/pdf/1708.02182.pdf">AWD LSTMs</a>, and to techniques such as <a href="https://arxiv.org/abs/1506.01186">cyclical learning rates</a> with random restarts. We extracted representations of phrases from this model by summarizing the hidden states using the concat pooling approach found in <a href="https://arxiv.org/pdf/1801.06146.pdf">this paper</a>.</p>
<p>One of the most challenging aspects of this exercise was to evaluate the quality of these embeddings. We are currently building a variety of downstream supervised tasks similar to those <a href="https://github.com/facebookresearch/SentEval">outlined here</a> that will aid us in evaluating the quality of these embeddings objectively. In the meantime, we sanity check our embeddings by manually examining the similarity between similar phrases. The below screenshot illustrates examples where we search the vectorized docstrings for similarity against user-supplied phrases:</p>
<p><img src="https://githubengineering.com/images/semantic_code_search/vec_sim.png" alt="vec sim"></p>
<h3 id="3-mapping-code-representations-to-the-same-vector-space-as-text">3. Mapping Code Representations To The Same Vector-Space as Text</h3>
<p>Next, we map the code representations we learned from the code summarization model (part 1) to the vector space of text. We accomplish this by fine-tuning the encoder of this model. The inputs to this model are still code blobs, however the target variable the model is now the vectorized version of docstrings. These docstrings are vectorized using the approach discussed in the previous section.</p>
<p>Concretely, we perform multi-dimensional regression with <a href="https://keras.io/losses/">cosine proximity loss</a> to bring the hidden state of the encoder into the same vector-space as text.</p>
<p>We are actively researching alternate approaches that directly learn a joint vector space of code and natural language, borrowing from some ideas outlined <a href="https://arxiv.org/abs/1705.00652">here</a>.</p>
<h3 id="4-creating-a-semantic-search-system">4. Creating a Semantic Search System</h3>
<p>Finally, after successfully creating a model that can vectorize code into the same vector-space as text, we can create a semantic search mechanism. In its most simple form, we can store the vectorized version of all code in a database, and perform nearest neighbor lookups to a vectorized search query.</p>
<p>Another active area of our research is determining the best way to augment existing keyword search with semantic results and how to incorporate additional information such as context and relevance. Furthermore, we are actively exploring ways to evaluate the quality of search results that will allow us to iterate quickly on this problem. We leave these topics for discussion in a future blog post.</p>
<h2 id="summary">Summary</h2>
<p>The below diagram summarizes all the steps in our current semantic-search workflow:</p>
<p><img src="https://githubengineering.com/images/semantic_code_search/overview.png" alt="code summarizer"></p>
<p>We are exploring ways to improve almost every component of this approach, including data preparation, model architecture, evaluation procedures, and overall system design. What is described in this blog post is only a minimal example that scratches the surface.</p>
<h2 id="open-source-examples">Open Source Examples</h2>
<p>Our <a href="https://towardsdatascience.com/semantic-code-search-3cd6d244a39c">open-source end-to-end tutorial</a> contains a detailed walkthrough of the approach outlined in this blog, along with code and data you can use to reproduce the results.</p>
<p>This open source example (with some modifications) is also used as a tutorial for the <a href="https://www.kubeflow.org/">kubeflow</a> project, which is implemented <a href="https://github.com/kubeflow/examples/tree/master/code_search">here</a>.</p>
<h3 id="limitations-and-intended-use-cases">Limitations and Intended Use Case(s)</h3>
<p>We believe that semantic code search will be most useful for targeted searches of code within specific entities such as repos, organizations or users as opposed to general purpose &ldquo;how to&rdquo; queries. The <a href="https://experiments.github.com/semantic-code-search">live demonstration</a> of semantic code search hosted on our <a href="https://blog.github.com/2018-09-18-introducing-experiments-an-ongoing-research-effort-from-github/">recently announced Experiments site</a> does not allow users to perform targeted searches of repos. Instead, this demonstration is designed to share a taste of what might be possible and searches only a limited, static set of python code.</p>
<p>Furthermore, like all machine learning techniques, the efficacy of this approach is limited by the training data used. For example, the data used to train these models are (code, docstring) pairs. Therefore, search queries that closely resemble a docstring have the greatest chance of success. On the other hand, queries that do not resemble a docstring or contain concepts for which there is little data may not yield sensible results. Therefore, it is not difficult to challenge our <a href="https://experiments.github.com/semantic-code-search">live demonstration</a> and discover the limitations of this approach. Nevertheless, our initial results indicate that this is an extremely fruitful area of research that we are excited to share with you.</p>
<p>There are many more use cases for semantic code search. For example, we could extend the ideas presented here to allow users to search for code using the language of their choice (French, Mandarin, Arabic, etc.) across many different programming languages simultaneously.</p>
<h2 id="get-in-touch">Get In Touch</h2>
<p>This is an exciting time for the machine learning research team at GitHub and we are looking to expand. If our work interests you, please <a href="https://github.com/about/careers">get in touch</a>!</p>
</div>
</div>
<br>
<br>
<br>
<br>
&#25216;&#26415;<br>
<br>
via GitHub Engineering https://ift.tt/2iXnnee<br>
<br>
September 19, 2018 at 01:51AM
